File: bigseqkit/faidx.go
package bigseqkit

import (
	"ignis/driver/api"
)

type SeqKitFaidxOptions struct {
	inner FaidxOptions
}

type FaidxOptions struct {
	Config     KitConfig
	UseRegexp  *bool
	IgnoreCase *bool
	FullHead   *bool
	RegionFile *string
	Regions    *[]string
}

func (this *FaidxOptions) setDefaults() *FaidxOptions {
	this.Config.setDefaults()
	setDefault(&this.UseRegexp, false)
	setDefault(&this.IgnoreCase, false)
	setDefault(&this.FullHead, false)
	setDefault(&this.RegionFile, "")
	setDefault(&this.Regions, make([]string, 0))

	return this
}

func (this *SeqKitFaidxOptions) Config(v *SeqKitConfig) *SeqKitFaidxOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitFaidxOptions) UseRegexp(v bool) *SeqKitFaidxOptions {
	this.inner.UseRegexp = &v
	return this
}

func (this *SeqKitFaidxOptions) IgnoreCase(v bool) *SeqKitFaidxOptions {
	this.inner.IgnoreCase = &v
	return this
}

func (this *SeqKitFaidxOptions) FullHead(v bool) *SeqKitFaidxOptions {
	this.inner.FullHead = &v
	return this
}

func (this *SeqKitFaidxOptions) RegionFile(v string) *SeqKitFaidxOptions {
	this.inner.RegionFile = &v
	return this
}

func (this *SeqKitFaidxOptions) Regions(v []string) *SeqKitFaidxOptions {
	this.inner.Regions = &v
	return this
}

func Faidx(input *api.IDataFrame[string], o *SeqKitFaidxOptions) (
	idx *api.IDataFrame[string], queries *api.IDataFrame[string], err error) {
	if o == nil {
		o = &SeqKitFaidxOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	offsets, err := api.MapPartitions[string, int64](input, libSource("FaidxOffset"))
	if err != nil {
		return nil, nil, err
	}
	offsetsArray, err := offsets.Collect()
	if err != nil {
		return nil, nil, err
	}
	for i := 0; i < len(offsetsArray)-1; i++ {
		offsetsArray[i+1] = offsetsArray[i]
	}
	offsetsArray[0] = 0

	libfaidx, err := api.AddParam(libSource("Faidx"), "offsets", offsetsArray)
	if err != nil {
		return nil, nil, err
	}

	libfaidx, err = api.AddParam(libfaidx, "opts", OptionsToString(opts))
	if err != nil {
		return nil, nil, err
	}

	faidx, err := api.MapPartitionsWithIndex[string, string](input, libfaidx)
	if err != nil {
		return nil, nil, err
	}

	if len(*opts.Regions) == 0 && *opts.RegionFile == "" {
		return faidx, nil, err
	}

	libqueries, err := api.AddParam(libSource("FaidxQuery"), "opts", OptionsToString(opts))
	if err != nil {
		return faidx, nil, err
	}

	queries, err = api.MapPartitions[string, string](input, libqueries)
	return faidx, queries, err
}

File: bigseqkit-py/bigseqkit/faidx.py
from typing import List

from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitFaidxOptions:

    def __init__(self):
        self.__inner = SeqOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def useRegexp(self, v: bool):
        self.__inner.UseRegexp = v

    def ignoreCase(self, v: bool):
        self.__inner.IgnoreCase = v

    def fullHead(self, v: bool):
        self.__inner.FullHead = v

    def regionFile(self, v: str):
        self.__inner.RegionFile = v

    def regions(self, v: List[str]):
        self.__inner.Regions = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()

        offsets = input.mapPartitions(_libSource("FaidxOffset"))
        offsetsArray = offsets.collect()
        for i in range(len(offsetsArray) - 1):
            offsetsArray[i + 1] = offsetsArray[i]
        offsetsArray[0] = 0

        libfaidx = _libSource("Faidx")\
            .addParam("offsets", offsetsArray)\
            .addParam("opts", _optionsToString(opts))

        faidx = input.mapPartitionsWithIndex(libfaidx)
        if len(opts.Regions) == 0 and opts.RegionFile == "":
            return faidx

        libqueries = _libSource("FaidxQuery").addParam("opts", _optionsToString(opts))

        return faidx, input.mapPartitions(libqueries)

class SeqOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.UseRegexp = None  # bool
        self.IgnoreCase = None  # bool
        self.FullHead = None  # bool
        self.RegionFile = None  # str
        self.Regions = None  # list[str]

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "UseRegexp", False)
        _setDefault(self, "IgnoreCase", False)
        _setDefault(self, "FullHead", False)
        _setDefault(self, "RegionFile", "")
        _setDefault(self, "Regions", [])


def faidx(input: IDataFrame, o: SeqKitFaidxOptions = None, **kwargs):
    if o is None:
        o = SeqKitFaidxOptions()
    return o._run(input, **kwargs)
